Remove noise from core tools log messages

Reduce noise from logging by caching the Assertion error

```
16:09:33  25/09/24 21:09:33 ERROR EventUtils: Unknown exception while parsing an event
16:09:33  java.lang.AssertionError: assertion failed: Expected hostname or IPv6 IP enclosed in [] 
16:09:33  	at scala.Predef$.assert(Predef.scala:223)
16:09:33  	at org.apache.spark.util.Utils$.checkHost(Utils.scala:1058)
16:09:33  	at org.apache.spark.storage.BlockManagerId.<init>(BlockManagerId.scala:50)
16:09:33  	at org.apache.spark.storage.BlockManagerId$.apply(BlockManagerId.scala:127)
16:09:33  	at org.apache.spark.util.JsonProtocol$.blockManagerIdFromJson(JsonProtocol.scala:1394)
16:09:33  	at org.apache.spark.util.JsonProtocol$.blockManagerAddedFromJson(JsonProtocol.scala:1011)
16:09:33  	at org.apache.spark.util.JsonProtocol$.sparkEventFromJson(JsonProtocol.scala:865)
16:09:33  	at org.apache.spark.util.JsonProtocol$.sparkEventFromJson(JsonProtocol.scala:850)
16:09:33  	at org.apache.spark.util.JsonProtocol.sparkEventFromJson(JsonProtocol.scala)
16:09:33  	at jdk.internal.reflect.GeneratedMethodAccessor14.invoke(Unknown Source)
16:09:33  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
16:09:33  	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
16:09:33  	at org.apache.spark.sql.rapids.tool.util.EventUtils$.$anonfun$runtimeEventFromJsonMethod$3(EventUtils.scala:204)
16:09:33  	at org.apache.spark.sql.rapids.tool.util.EventUtils$.$anonfun$getEventFromJsonMethod$2(EventUtils.scala:215)
16:09:33  	at scala.util.Try$.apply(Try.scala:213)
16:09:33  	at org.apache.spark.sql.rapids.tool.util.EventUtils$.$anonfun$getEventFromJsonMethod$1(EventUtils.scala:215)
16:09:33  	at org.apache.spark.sql.rapids.tool.AppBase.$anonfun$processEventsInternal$7(AppBase.scala:350)
16:09:33  	at org.apache.spark.sql.rapids.tool.AppBase.$anonfun$processEventsInternal$7$adapted(AppBase.scala:346)
16:09:33  	at scala.collection.Iterator.find(Iterator.scala:995)
16:09:33  	at scala.collection.Iterator.find$(Iterator.scala:992)
16:09:33  	at scala.collection.AbstractIterator.find(Iterator.scala:1431)
16:09:33  	at org.apache.spark.sql.rapids.tool.AppBase.$anonfun$processEventsInternal$5(AppBase.scala:346)
16:09:33  	at org.apache.spark.util.SparkErrorUtils.tryWithResource(SparkErrorUtils.scala:48)
16:09:33  	at org.apache.spark.util.SparkErrorUtils.tryWithResource$(SparkErrorUtils.scala:46)
16:09:33  	at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:94)
16:09:33  	at org.apache.spark.sql.rapids.tool.AppBase.$anonfun$processEventsInternal$3(AppBase.scala:345)
16:09:33  	at scala.collection.immutable.List.foreach(List.scala:431)
16:09:33  	at org.apache.spark.sql.rapids.tool.AppBase.processEventsInternal(AppBase.scala:344)
16:09:33  	at org.apache.spark.sql.rapids.tool.AppBase.processEvents(AppBase.scala:556)
16:09:33  	at org.apache.spark.sql.rapids.tool.profiling.ApplicationInfo.<init>(ApplicationInfo.scala:196)
16:09:33  	at com.nvidia.spark.rapids.tool.profiling.Profiler.createApp(Profiler.scala:215)
16:09:33  	at com.nvidia.spark.rapids.tool.profiling.Profiler.com$nvidia$spark$rapids$tool$profiling$Profiler$$profileApp(Profiler.scala:140)
16:09:33  	at com.nvidia.spark.rapids.tool.profiling.Profiler$ProfileProcessThread$1.run(Profiler.scala:180)
16:09:33  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
16:09:33  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
16:09:33  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
16:09:33  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
16:09:33  	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/core/src/main/scala/com/nvidia/spark/rapids/tool/EventLogPathProcessor.scala
+++ b/core/src/main/scala/com/nvidia/spark/rapids/tool/EventLogPathProcessor.scala
@@ -449,7 +449,7 @@ object EventLogPathProcessor extends Logging {
   }
 
   def logApplicationInfo(app: ApplicationInfo) = {
-    logInfo(s"==============  ${app.appId} ==============")
+    logDebug(s"==============  ${app.appId} ==============")
   }
 
   def getDBEventLogFileDate(eventLogFileName: String): LocalDateTime = {

--- a/core/src/main/scala/com/nvidia/spark/rapids/tool/Platform.scala
+++ b/core/src/main/scala/com/nvidia/spark/rapids/tool/Platform.scala
@@ -943,7 +943,7 @@ class OnPremPlatform(gpuDevice: Option[GpuDevice],
           gpuDevice = gpuDevice)
       }
     }.orElse {
-      logInfo("Worker info or Gpu info is not provided in the target cluster. " +
+      logDebug("Worker info or Gpu info is not provided in the target cluster. " +
         "Skipping recommended instance info creation.")
       None
     }

--- a/core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Profiler.scala
+++ b/core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Profiler.scala
@@ -218,7 +218,7 @@ class Profiler(hadoopConf: Configuration, appArgs: ProfileArgs, enablePB: Boolea
       if (!app.isAppMetaDefined) {
         throw IncorrectAppStatusException()
       }
-      logInfo(s"Took ${endTime - startTime}ms to create App for ${path.eventLog.toString}")
+      logDebug(s"Took ${endTime - startTime}ms to create App for ${path.eventLog.toString}")
       Right(app)
     } catch {
       case e: Exception =>
@@ -277,7 +277,7 @@ class Profiler(hadoopConf: Configuration, appArgs: ProfileArgs, enablePB: Boolea
     val endTime = System.currentTimeMillis()
     val appInfo = collect.getAppInfo
     val sqlMetrics = collect.getSQLPlanMetrics
-    logInfo(s"Time to collect Profiling Info [${appInfo.head.appId}]: ${endTime - startTime}.")
+    logDebug(s"Time to collect Profiling Info [${appInfo.head.appId}]: ${endTime - startTime}.")
     val appInfoSummary = ApplicationSummaryInfo(
       appInfo = appInfo,
       dsInfo = collect.getDataSourceInfo(sqlMetrics),

--- a/core/src/main/scala/org/apache/spark/sql/rapids/tool/util/EventUtils.scala
+++ b/core/src/main/scala/org/apache/spark/sql/rapids/tool/util/EventUtils.scala
@@ -39,10 +39,46 @@ object EventUtils extends Logging {
   // Set to keep track of missing classes
   private val missingEventClasses = mutable.HashSet[String]()
 
+  // This is the prefix of the assertion error message we want to track when IPV6 is not valid.
+  private val IPV6_ASSERTION_ERR_PREFIX =
+    "assertion failed: Expected hostname or IPv6 IP enclosed in [] but got"
+  // A cache to avoid reporting the same unknown exception more than once
+  // The key is the exception message, and the value is the count of occurences
+  private val cachedUnknownExceptions = new java.util.concurrent.ConcurrentHashMap[String, Int]()
+
   private def reportMissingEventClass(className: String): Unit = {
     if (!missingEventClasses.contains(className)) {
       missingEventClasses.add(className)
       logWarning(s"ClassNotFoundException while parsing an event: $className")
+    }
+  }
+
+  /**
+   * Handles AssertionError exceptions, specifically to catch and log the
+   * "assertion failed: expected hostname or IPv6" error only once. If the error keeps occuring we
+   * only report it once.
+   * @param t the Throwable to handle
+   */
+  private def handleAssertionError(t: Throwable): Unit = {
+    // report assertion failed: expected hostname or IPv6 only once
+    val (keyErrMsg, dumpStack) = if (t.getMessage.startsWith(IPV6_ASSERTION_ERR_PREFIX)) {
+      // use a constant key to track this specific error and suppress the entire stack.
+      (IPV6_ASSERTION_ERR_PREFIX, false)
+    } else {
+      (t.getMessage, true)
+    }
+    val errCount =
+      cachedUnknownExceptions.compute(keyErrMsg, (_, v) => Option(v).map(_ + 1).getOrElse(1))
+    if (errCount == 1) {
+      // only report the first occurrence
+      val errMessage = "Assertion Error encountered while parsing an event. " +
+        "Only the first occurrence will be logged; subsequent similar errors will be suppressed " +
+        "for clarity.\n\t\t"
+      if (dumpStack) {
+        logError(errMessage, t)
+      } else {
+        logError(errMessage + s"${t.getMessage}")
+      }
     }
   }
 
@@ -237,6 +273,10 @@ object EventUtils extends Logging {
                 case z: ClassNotFoundException if z.getMessage != null =>
                   // Avoid reporting missing classes more than once to reduce noise in the logs
                   reportMissingEventClass(z.getMessage)
+                case a: AssertionError =>
+                  // This is to catch "assertion failed: expected hostname or IPv6"
+                  // which may be caused by malformed eventlog.
+                  handleAssertionError(a)
                 case t: Throwable =>
                   // We do not want to swallow unknown exceptions so that we can handle later
                   logError(s"Unknown exception while parsing an event", t)

